azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py

```diff
@@ -232,8 +236,96 @@ def _parse_agent_manifests(self, vm_settings):
+        #             "version": "1.2.0",
+        #             "location": "https://umsavbvncrpzbnxmxzmr.blob.core.windows.net/f4086d41-69f9-3103-78e0-8a2c7e789d0f/f4086d41-69f9-3103-78e0-8a2c7e789d0f_manifest.xml",
+        #             "failoverlocation": "https://umsajbjtqrb3zqjvgb2z.blob.core.windows.net/f4086d41-69f9-3103-78e0-8a2c7e789d0f/f4086d41-69f9-3103-78e0-8a2c7e789d0f_manifest.xml",
+        #             "additionalLocations": [
+        #                 "https://umsawqtlsshtn5v2nfgh.blob.core.windows.net/f4086d41-69f9-3103-78e0-8a2c7e789d0f/f4086d41-69f9-3103-78e0-8a2c7e789d0f_manifest.xml"
+        #             ],
+        #             "state": "enabled",
+        #             "autoUpgrade": true,
+        #             "runAsStartupTask": false,
+        #             "isJson": true,
+        #             "useExactVersion": true,
+        #             "settingsSeqNo": 0,
+        #             "isMultiConfig": true,
+        #             "settings": [
+        #                 {
+        #                     "publicSettings": "{\"source\":{\"script\":\"echo '4abb1e88-f349-41f8-8442-247d9fdfcac5'\"}}",
+        #                     "seqNo": 0,
+        #                     "extensionName": "MCExt1",
+        #                     "extensionState": "enabled"
+        #                 },
+        #                 {
+        #                     "publicSettings": "{\"source\":{\"script\":\"echo 'e865c9bc-a7b3-42c6-9a79-cfa98a1ee8b3'\"}}",
+        #                     "seqNo": 0,
+        #                     "extensionName": "MCExt2",
+        #                     "extensionState": "enabled"
+        #                 },
+        #                 {
+        #                     "publicSettings": "{\"source\":{\"script\":\"echo 'f923e416-0340-485c-9243-8b84fb9930c6'\"}}",
+        #                     "seqNo": 0,
+        #                     "extensionName": "MCExt3",
+        #                     "extensionState": "enabled"
+        #                 }
+        #             ]
+        #         }
+        #         ...
+        #     ]
+        #     ...
+        # }
+        extension_goal_states = vm_settings.get("extensionGoalStates")
+        if extension_goal_states is not None:
+            if not isinstance(extension_goal_states, list):
+                raise Exception("extension_goal_states should be an array")
+            for extension_gs in extension_goal_states:
+                extension = Extension()
+
+                extension.name = extension_gs['name']
+                extension.version = extension_gs['version']
+                extension.state = extension_gs['state']
+                if extension.state not in ExtensionRequestedState.All:
+                    raise Exception('Invalid extension state: {0} ({1})'.format(extension.state, extension.name))
+                is_multi_config = extension_gs.get('isMultiConfig')
```

Comment:
Maybe you can modify the `CaseFoldedDict.get()` function to accept a default parameter to avoid using `None` checks in all these places

Reply:
The CaseFoldedDict will go away eventually. but in any case I feel the explicit check for None is better for those optional properties. Creating custom dictionaries with default values for the different pieces of the goal state would obscure the code, I think.